공연 선점 기능

kochungcheon opened the pull request:
# 목표로 한 것

- 공연 선점 기능 개발 

- 공연 조회 로직에 공연 선점에 따른 변화 반영

# 고려한 것

- 레디스 키 값으로 유저와 티켓 id 를 넘겨주었습니다. 

- 해당 기능은 여러 도메인의 의존성이 필요하다 판단해 별도의 클래스로 분리하였습니다

kochungcheon committed:
feat: 좌석 선점 로직 구현

[좌석 선점 로직]

- SeatOccupiedController 의 occupiedSeat

- SeatOccupiedService 의 occupiedSeat

- StringRedisTemplates 를 통해 구현했습니다

[좌석 조회 로직]

- SeatOccupiedController 의 getAvailableSeats

- SeatOccupiedService 의 getAvailableSeats 와 getReservedCountBySeatId

[SeatOccupiedService 생성]

- 의존성 처리를 위헤 별도 서비스를 생성하였습니다

kochungcheon committed:
feat: 테스트 코드 추가

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedController.java:
```diff
@@ -0,0 +1,31 @@
+public class SeatOccupiedController {
+	private final SeatOccupiedService seatOccupiedService;
+
+	@PostMapping("/api/SeatOccupied")
```
오.. 처음보는 단어군요. 다만 uri는 리소스와 행동을 나타내기 때문에 과거형을 쓰기보다는 동사형태를 사용하는게 어떨까 싶군용? occupy 라는게 있군요

kochungcheon replied:
넵 알겠습니다! 맞춰 수정하겠습니다!

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedController.java:
```diff
@@ -0,0 +1,31 @@
+		seatOccupiedService.occupiedSeat(ticketId, userId, requestCount);
+	}
+
+	@GetMapping("/apiV2/shows/{performanceId}/availableSeat")
```
요것도 restAPI 스럽게 availableSeat의 복수를 쓰는게 어떨까 싶습니다.

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedService.java:
```diff
@@ -0,0 +1,77 @@
+import lombok.AllArgsConstructor;
+
+@Service
+@AllArgsConstructor
```
@RequiredArgsConstructor 가 아닌 @AllArgsConstructor를 사용함에 따라 private StringRedisTemplate redisTemplate만 final이 없어도 동작하는 군요. 통일하면 좋을 것 같습니다~

kochungcheon replied:
넵 통일 하였습니다!

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedService.java:
```diff
@@ -0,0 +1,77 @@
+	public void occupiedSeat(String ticketId, Long userId, int requestCount) {
+
+		String userKey = String.valueOf(userId);
+		Long remainingSeats = redisTemplate.opsForHash().increment(ticketId, userKey, -requestCount);
```
repository를 interface로 분리해 domain <- jpa repository 의 의존방향을 만든 것 처럼 
redisTemplate 또한 interface로 분리해 domain <- redis 로 만들 수 있을겁니다~

heeve1 replied:
그리고 레디스는 보통 서비스 내 여러 곳에서 캐싱용도로 사용되기 때문에 현제처럼 ticketId를 그대로 key로 사용하는 경우 다른 곳에서 또 레디스를 사용하기 힘들겁니다. 보통 nameSpace1:key1:nameSpace2:key2 같이 쓰곤 하지요. 
따라서 이경우엔 TICKET_ID:${ticketId} 같이도 써볼 수 있을 것 같네요. (포멧은 자유입니다)

kochungcheon replied:
레디스 키를 TICKET_ID:{} 형태로 변경하였습니다 그리고 인터페이스를 통해 의존 방향을 역전시켰습니다!

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedService.java:
```diff
@@ -0,0 +1,77 @@
+		// 각 seatId에 대한 선점 좌석 수 계산
+		Map<Grade, Integer> reservedSeats = new EnumMap<>(Grade.class);
+		for (Long ticketId : ticketList) {
+			Ticket ticket = ticketDataAccessService.findTicketBy(ticketId);
```
기본적으로 for 같은 반복문 안에서 findTicketBy 같은 쿼리를 요청하는 경우 JPA N+1과 유사한 패턴이 연출됩니다.
해당 코드는 performance에 포함된 ticket의 수만큼 쿼리가 발생되겠군요! 

보통 이 경우 for문 바깥에서 대상 엔티티를 한번에 조회한 후 for문 안에서 한개한개 매핑해 주곤합니다~

kochungcheon replied:
아하 그렇군요..! 한번에 불러오는 형식 (findTicketsBy) 해주었습니다!

heeve1 reviewed (commented):
고생하셨습니다~~~

kochungcheon committed:
refactor: PR 반영

- RESTFUL 하게 리소스 변경

- @requiredargsconstructor 변경

- 레디스 키 변경 TICKET_ID:{ticket_id}

- SeatOccupy 와 SeatOccupyImpl 생성 의존 역전

- findTicketsBy 생성

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedService.java:
```diff
@@ -15,50 +13,40 @@
 
 		String userKey = String.valueOf(userId);
-		Long remainingSeats = redisTemplate.opsForHash().increment(ticketId, userKey, -requestCount);
+		String redisKey = seatOccupy.makeRedisKey(ticketId);
```
DIP 를 통해서 중요한 코드가 기술 코드를 의존하지 않고자 하였는데 makeRedisKey는 기술과 관련된 이름인것 같군요 😭

heeve1 commented on src/main/java/com/ticketease/te/seatOccupied/SeatOccupiedService.java:
```diff
@@ -74,4 +62,15 @@ public GradeCount getAvailableSeats(Long performanceId) {
 		return GradeCount.of(availableSeats);
 	}
 
+	public int getReservedCountBySeatId(String ticketId) {
+		String redisKey = seatOccupy.makeRedisKey(ticketId);
+		Map<Object, Object> userReservations = seatOccupy.getHashEntries(redisKey);
+
+		int totalReserved = 0;
+		for (Object count : userReservations.values()) {
+			totalReserved += (Integer)count;
+		}
+
+		return totalReserved;
+	}
```
getReservedCountBySeatId는 굉장히 자주 호출될 것 같은데 성능이 걱정되는 포인트군요 ㅎㅎ 어떻게 하면 좋을지 추후 고민해보면 좋을 것 같아요

smh0300 reviewed (approved):
고생하셨습니다!
